### ISSUE TITLE
Redisson reconnection issue.

### ISSUE THREAD

**Issue by amitfou** [redisson/redisson#914]
When the network is unavailable I get the following error

org.redisson.client.RedisTimeoutException: Command execution timeout for xxxxxx/xx.xx.xx.xx:6379
	at org.redisson.client.RedisConnection$2.run(RedisConnection.java:193) [redisson-3.4.2.jar:?]
	at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38) [netty-common-4.1.10.Final.jar:4.1.10.Final]
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120) [netty-common-4.1.10.Final.jar:4.1.10.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) [netty-common-4.1.10.Final.jar:4.1.10.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403) [netty-common-4.1.10.Final.jar:4.1.10.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462) [netty-transport-4.1.10.Final.jar:4.1.10.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) [netty-common-4.1.10.Final.jar:4.1.10.Final]
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144) [netty-common-4.1.10.Final.jar:4.1.10.Final]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_73]

Which is expected but once the network connection is restored redisson never reconnects. I am using ElasticCache and Redisson version 3.4.2. Most of my settings are default, I tried some settings but nothing works. I presume that this should happen by default which is the case when I use Jedis

**Comment by mrniko** [comment]
> Which is expected but once the network connection is restored redisson never reconnects

Redisson always resends a command once it has been reconnected to Redis. 
Could you please show your settings?

**Comment by amitfou** [comment]
it resends the command but never reconnects. My setup is very simple as I am still debugging this issue at the very least I am using.

 config.useSingleServer()
                    .setAddress(redisUrl + ":" + redisPort)
                    .setConnectTimeout(5000);

I also tried using

config.useReplicatedServers()
                    .setScanInterval(2000)
                    .addNodeAddress(primary + ":" + redisPort)
                    .addNodeAddress(replica + ":" + redisPort)
                    .setReadMode(ReadMode.SLAVE)
                    .setReconnectionTimeout(1500)
                    .setFailedAttempts(Integer.MAX_VALUE)
                    .setConnectTimeout(5000);

It would be good to have a section describing how to connect to ElasticCache.

**Comment by mrniko** [comment]
Each Redisson connection has `org.redisson.client.handler.ConnectionWatchdog` handler. It reconnects connection each time when connection has been lost.

I have added test for this case: `RedissonTest.testReconnection`.

**Comment by amitfou** [comment]
Thanks Nikita, but the issue is that after the first disconnection I keep getting the following even after the connection has been restored (just to confirm the Jedis api works fine in this scenario). 

org.redisson.client.RedisTimeoutException: Redis server response timeout (5000 ms) occured for command: (GET) with params: [amit:userConversationState:62385260-1fba-4d4c-b508-608bde66a798] channel: [id: 0x4cad2ba7, L:/x.x.x.143:57717 - R:xxxxxxxx.cache.amazonaws.com/xx.x.x.123:6379]
	at org.redisson.command.CommandAsyncService$10.run(CommandAsyncService.java:647) ~[redisson-3.4.2.jar:?]
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:661) ~[netty-common-4.1.10.Final.jar:4.1.10.Final]
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:736) ~[netty-common-4.1.10.Final.jar:4.1.10.Final]
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:464) ~[netty-common-4.1.10.Final.jar:4.1.10.Final]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_73]

I understand its something to do with my setup as I am in Sydney and connecting to redis instance in Singapore aws. I am not having any issues in my EC2 server as the scenario doesn't happen their but it happens on my local development machine. There is for sure some issue and happens because of my setup but I would consider it as a bug as it doesn't happen when I use other apis. May be I will try to debug more and give you some more info on this later.

Thanks again for looking into the issue.

**Comment by JathinSanghvi** [comment]
did you get this working. I am having similar issues. Any help will be appreciated.

**Comment by mrniko** [comment]
Unable to reproduce it. Here is my code:

```java
        Config config = new Config();
        config.useSingleServer().setAddress("redis://127.0.0.1:6399");

        RedissonClient r = Redisson.create(config);
        for (int i = 0; i < 10000; i++) {
            try {
                r.getMap("test").put(i, i);
                r.getMap("test").get(i);
            } catch (Exception e) {
                e.printStackTrace();
            }
            Thread.sleep(100);
            System.out.println(i);
        }
```

Log is attached
[log.txt](https://github.com/redisson/redisson/files/1138707/log.txt)

**Comment by wintersliu** [comment]
I am having same issue against 3.5.4 version. Not sure if this only happens using AWS ElastiCache backend. @mrniko Do you mind testing against ElastiCache?

**Comment by mrniko** [comment]
Did you try to set `pingConnectionInterval` setting? it might help

**Comment by mrniko** [comment]
Fixed in 3.6.2 version
